lib/lf/refinement/mesh_hierarchy.h

```diff
@@ -101,8 +101,8 @@ class MeshHierarchy {
    * created by refinement.
    * - Unmarks all edges, see RefineMarked().
    */
-  MeshHierarchy(std::shared_ptr<mesh::Mesh> base_mesh,
-                std::shared_ptr<mesh::MeshFactory> mesh_factory);
+  MeshHierarchy(const std::shared_ptr<mesh::Mesh> &base_mesh,
```

Comment:
A mesh hierarchy should own a mutable shared pointer to a mesh. Just imagine, a mesh hierarchy is built based on some coarse mesh. Then the shared pointer to the coarse mesh reaches the end of its scope. The coarse mesh must not be deleted, lest the mesh hierarchy lose one of its levels. In my opinion this rules out using a constant shared pointer. I am surprised that the compiler does not complain.

Reply:
The mesh hierarchy still owns a shared_ptr to the base_mesh. If you look at Line 36 of https://github.com/craffael/lehrfempp/blob/issue_133/lib/lf/refinement/mesh_hierarchy.cc you can see that a copy of `base_mesh` is pushed onto `meshes_` which is a private member field of the class MeshHierarchy. So it's not possible that the coarsest mesh of a MeshHierarchy gets deleted before the MeshHierarchy itself.

The change from `std::shared_ptr<mesh::Mesh>` to `const std::shared_ptr<mesh::Mesh>&` was actually encouraged by clang-tidy. By doing so we save one unnecessary copy of shared_ptr. I would have to check why clang-tidy didn't suggest this already before.